client/client.ml

```diff
@@ -49,10 +48,16 @@ let send_msg ~profilename ~confdir ~conffile msg =
   let uri = Uri.make ~scheme:"http" ~host:hostname ~port () in
   let prefix = Oca_lib.protocol_version^"\n"^prefix in
   print_endline "Sending command…";
-  Lwt_main.run begin
-    let%lwt resp = Cohttp_lwt_unix.Client.post ~body:(`String (prefix^msg)) uri in
-    process_response resp
-  end
+  match
+    Httpcats.request ~meth:`POST ~body:(Httpcats.String (prefix^msg))
+      ~uri:(Uri.to_string uri)
+      ~f:(fun _ res () body ->
+        match res.Httpcats.status, body with
+        | (`OK | `Upgrade_required), Some x -> print_partial_body x
+        | _, None | _, _ -> ()) ()
```

Comment:
Compared to `Cohttp_lwt_unix.Client.post`, `Httpcats.request ~meth:`POST` feels a bit clanky around its fold-style function. I helped create that API (https://github.com/robur-coop/http-lwt-client/pull/14/files#diff-ea23e09b315e06abe1b96d61b5ea98f62833a893a5e7f238af91b7c8d47adc40R71) but at the time the `response` wasn't part of the parameters. Now i see the need to have the response around but it feels clanky to check the response several times. I think we can have a better API that feels less clanky to use.

Do you have any idea?

Reply:
Unfortunately, the main problem concerns redirects. This is something that should be documented but, currently, httpcats can call `f` several times with several responses (the responses with the redirection as well as the content of these responses and the final response with its content).

An example of how to handle this is available here: https://github.com/robur-coop/hurl/blob/db73f3fe964a1549ca6acef8f720d55e5454faa0/bin/hurl.ml#L79-L92

It's clearly not satisfactory but I'd need a lot of iteration to find a good API.